README.md
# Visible Wordpress Starter

A Docker Wordpress development environment by the team at
[Visible](https://visible.vc/) and some awesome
[contributors](https://github.com/visiblevc/wordpress-starter/graphs/contributors).
Our goal is to make Wordpress development slightly less frustrating.

-   [Introduction](#introduction)
-   [Example](./example/)
-   [Requirements](#requirements)
-   [Getting Started](#getting-started)
-   [Available Images](#available-images)
-   [Default Wordpress Admin Credentials](#default-wordpress-admin-credentials)
-   [Default Database Credentials](#default-database-credentials)
-   [Service Environment Variables](#service-environment-variables)
    -   [`wordpress`](#wordpress)
    -   [`db`](#db)
-   [Workflow Tips](#workflow-tips)
    -   [Using `wp-cli`](#using-wp-cli)
    -   [Working with Databases](#working-with-databases)
-   [Using in Production](#using-in-production)
-   [Contributing](#contributing)

### Introduction

We wrote a series of articles explaining in depth the philosophy behind this
project:

-   [Intro: A slightly less shitty WordPress developer workflow](https://visible.vc/engineering/wordpress-developer-workflow/)
-   [Part 1: Setup a local development environment for WordPress with Docker](https://visible.vc/engineering/docker-environment-for-wordpress/)
-   [Part 2: Setup an asset pipeline for WordPress theme development](https://visible.vc/engineering/asset-pipeline-for-wordpress-theme-development/)
-   [Part 3: Optimize your wordpress theme assets and deploy to S3](https://visible.vc/engineering/optimize-wordpress-theme-assets-and-deploy-to-s3-cloudfront/)

### Requirements

Well, to run a Docker environment, you will need Docker. The Dockerfile is only
for an Apache+PHP+Wordpress container, you will need a `MySQL` or `MariaDB`
container to run a website. We use Docker Compose 1.6+ for the orchestration.

### Getting started

This project has 2 parts: the Docker environment and a set of tools for theme
development. To quickly get started, you can simply run the following:

```
# copy the files
git clone https://github.com/visiblevc/wordpress-starter.git

# navigate to example directory
cd wordpress-starter/example

# start the website at localhost:8080
docker-compose up -d && docker-compose logs -f wordpress
```

**NOTE:** If you run on MacOS with Docker in VirtualBox, you will want to
forward the port by running this
`VBoxManage controlvm vm-name natpf1 "tcp8080,tcp,127.0.0.1,8080,,8080"`. If you
use another port than `8080`, change it in the command.

### Available Images

| PHP Version | Tags                                        |
| ----------- | ------------------------------------------- |
| **7.4**     | `latest` `latest-php7.4` `<version>-php7.4` |
| **7.3**     | `latest` `latest-php7.3` `<version>-php7.3` |
| **7.2**     | `latest-php7.2` `<version>-php7.2`          |

If you need a specific version, look at the [Changelog](CHANGELOG.md)

### Default Wordpress Admin Credentials

To access the Wordpress Admin at `/wp-admin`, the default values are as follows:

| Credential            | Value                           | Notes                                                      |
| --------------------- | ------------------------------- | ---------------------------------------------------------- |
| **Username or Email** | `root` or `admin@wordpress.com` | Can be changed with the `ADMIN_EMAIL` environment variable |
| **Password**          | `root`                          | Uses the same value as the `DB_PASS` environment variable  |

### Default Database Credentials

| Credential        | Value                  | Notes                                                                                              |
| ----------------- | ---------------------- | -------------------------------------------------------------------------------------------------- |
| **Hostname**      | `db`                   | Can be changed with the `DB_HOST` environment variable **NOTE:**: Must match database service name |
| **Username**      | `root`                 |                                                                                                    |
| **Password**      |                        | Must be set using the `DB_PASS` environment variable                                               |
| **Database Name** | `wordpress`            | Can be changed with the `DB_NAME` environment variable                                             |
| **Admin Email**   | `admin@${DB_NAME}.com` |                                                                                                    |

### Service Environment Variables

**Notes:**

-   Variables marked with ✅ are required
-   Single quotes must surround `boolean` environment variables

#### `wordpress`

| Variable           | Default Value                    | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
| ------------------ | -------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `DB_USER`          | `root`                           | Username for both the database and the WordPress installation (if not importing existing)                                                                                                                                                                                                                                                                                                                                                                                                                                               |
| `DB_PASS`✅        |                                  | Password for the database. Value must match `MYSQL_ROOT_PASSWORD` set in the `db` service                                                                                                                                                                                                                                                                                                                                                                                                                                               |
| `DB_HOST`          | `db`                             | Hostname for the database                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
| `DB_NAME`          | `wordpress`                      | Name of the database                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| `DB_PREFIX`        | `wp_`                            | Prefix for the database                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| `DB_CHARSET`       | `utf8`                           | Select a charset for the wordpress database (legacy versions might not be utf8)                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| `SERVER_NAME`      | `localhost`                      | Set this to `<your-domain-name>.<your-top-level-domain>` if you plan on obtaining SSL certificates                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| `ADMIN_EMAIL`      | `admin@${DB_NAME}.com`           | Administrator email address                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
| `WP_LOCALE`        | `en_US`                          | Set the site language                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| `WP_DEBUG`         | `'false'`                        | [Click here](https://codex.wordpress.org/WP_DEBUG) for more information                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| `WP_DEBUG_DISPLAY` | `'false'`                        | [Click here](https://codex.wordpress.org/WP_DEBUG#WP_DEBUG_DISPLAY) for more information                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| `WP_DEBUG_LOG`     | `'false'`                        | [Click here](https://codex.wordpress.org/WP_DEBUG#WP_DEBUG_LOG) for more information                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| `WP_VERSION`       | `latest`                         | Specify the WordPress version to install. Accepts any valid semver number, `latest`, or `nightly` for beta builds.                                                                                                                                                                                                                                                                                                                                                                                                                      |
| `THEMES`           |                                  | Space-separated list of themes you want to install in either of the following forms<ul><li>`theme-slug`: Used when installing theme direct from WordPress.org</li><li>`[theme-slug]https://themesite.com/theme.zip`: Used when installing theme from URL</li></ul>                                                                                                                                                                                                                                                                      |
| `PLUGINS`          |                                  | Space-separated list of plugins you want to install in either of the following forms:<ul><li>`plugin-slug`: Used when installing plugin direct from WordPress.org.</li><li>`[plugin-slug]http://pluginsite.com/plugin.zip`: Used when installing plugin from URL.</li></ul>                                                                                                                                                                                                                                                             |
| `MULTISITE`        | `'false'`                        | Set to `'true'` to enable multisite                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| `PERMALINKS`       | `/%year%/%monthnum%/%postname%/` | A valid WordPress permalink [structure tag](https://codex.wordpress.org/Using_Permalinks#Structure_Tags)                                                                                                                                                                                                                                                                                                                                                                                                                                |
| `URL_REPLACE`      |                                  | <p>Value must be a full replacement URL to use in development environments if importing a database for development use.</p><p><strong>Example:</strong> If your live site's URL is `https://www.example.com` and you intend to use the database on port 8080 of localhost, this value should be set to `http://localhost:8080`.</p><p><strong>Note:</strong> If you are running Docker using Docker Machine, your replacement url MUST be the output of the following command: `echo $(docker-machine ip <your-machine-name>):8080`</p> |
| `EXTRA_PHP`        |                                  | <p>Extra PHP code to add to `wp-config.php`. [Click here](https://developer.wordpress.org/cli/commands/config/create/) for more information.</p><p><strong>IMPORTANT NOTE:</strong> All `$` symbols must be escaped by prepending an extra `$`, otherwise it will be interpreted by docker as an environment variable. In other words, `$variable` must be `$$variable`.</p>                                                                                                                                                                                                                       |

#### `db`

| Variable                | Default Value | Description                                     |
| ----------------------- | ------------- | ----------------------------------------------- |
| `MYSQL_ROOT_PASSWORD`✅ |               | Must match `DB_PASS` of the `wordpress` service |

## Workflow Tips

### Using `wp-cli`

You can access wp-cli by running `npm run wp ...`. Here are some examples:

```
npm run wp plugin install <some-plugin>
npm run wp db import /data/database.sql
```

### Working with Databases

If you have an exported `.sql` file from an existing website, drop the file into
the `data/` folder. The first time you run the container, it will detect the SQL
dump and use it as a database. If it doesn't find one, it will create a fresh
database.

If the SQL dump changes for some reason, you can reload the database by running:

```sh
docker-compose exec wordpress /bin/bash "wp db import $(find /data/*.sql | head -n 1) --allow-root"
```

If you want to create a dump of your development database, you can run:

```sh
docker-compose exec wordpress /bin/bash -c 'wp db export /data/dump.sql --allow-root'
```

Finally, sometimes your development environment runs on a different domain than
your live one. The live will be `example.com` and the development
`localhost:8080`. This project does a search and replace for you. You can set
the `URL_REPLACE: localhost:8080` environment variable in the
`docker-compose.yml`.

## Using in Production

### Adjustments to `docker-compose.yml`

```yml
# If something isn't shown, assume it's the same as the examples above
version: "3"
services:
    wordpress:
        ports:
            - 80:80
            - 443:443
        restart: always
        environment:
            SERVER_NAME: mysite.com
            DB_PASS: ${SECURE_PASSWORD} # Stored in .env file
        volumes:
            - ./letsencrypt:/etc/letsencrypt
            - ./data:/data
            # anything else you'd like to be able to back up
    db:
        restart: always
        environment:
            MYSQL_ROOT_PASSWORD: ${SECURE_PASSWORD} # Stored in .env file
```

### SSL Certificates

We highly recommend securing your site with SSL encryption. The Let's Encrypt
and Certbot projects have made doing this both free (as in beer) and painless.
We've incorporated these projects into this project.

Assuming your site is running on your production host, follow the below steps to
obtain and renew SSL certificates.

#### Obtaining Certificates

You should first [set `SERVER_NAME` to `<your-domain-name>.<your-top-level-domain>` in your `docker-compose.yml`](#wordpress)

```sh
$ docker-compose ps
Name                   Command                        State
---------------------------------------------------------
project_db_1           docker-entrypoint.sh mysqld     Up
project_wordpress_1    docker-php-entrypoint /run.sh   Up

$ docker-compose exec wordpress /bin/bash
root@4e16c7fe4a10:/app# certbot --apache
```

#### Renewing Certificates

```sh
$ docker-compose ps
Name                   Command                        State
---------------------------------------------------------
project_db_1           docker-entrypoint.sh mysqld     Up
project_wordpress_1    docker-php-entrypoint /run.sh   Up

$ docker-compose exec wordpress /bin/bash
root@4e16c7fe4a10:/app# certbot renew
```

## Contributing

You can find Development instructions in the
[Wiki](https://github.com/visiblevc/wordpress-starter/wiki/Development).


CHANGELOG.md
# Changelog

## 0.25.4 - latest

### Patch

- Fix bug that prevented url form of plugins/themes to fail due to nullglob interacting with the parsing. #176

## 0.25.3

### Patch

- Fix bug preventing a default theme from being installed when no theme dependencies or volumes exist. #172
- Fix small bug that occurred when no volumes were included in `/data` due to not having nullglob set.

## 0.25.2

### Patch

- Fix bug in plugin/theme activation caused by attempting to activate using a URL rather than the slug in cases where plugins/themes are installed using zip URL.

## 0.25.1

### Minor

- Drop PHP 7.0 and 7.1 images.
- Add PHP 7.4 image.

### Patch

- Fix bug in run.sh resulting from an invalid flag given to `find`

## 0.25.0

### Minor

- Make runtime script much simpler and more reliable. #165 
- Increase verbosity of logging related to database interactions. #165 

## 0.24.0

### Minor

* feat: bump default theme install to "twentytwenty"
* feat: increase `upload_max_filesize` to `50M`

### Patch

* fix: include "must-use" plugins in plugin volume check
* fix: add --create-dirs to curl in dockerfile
* fix: improve plugin volumes check to also include plugins that are a single php file

## 0.23.0

### Minor 

- Add missing `bcmath` and `imagick` PHP extensions recommended by WordPress 5.2's Health Check.

## 0.22.0

### Minor

- Add `php7.3`
- Add `/docker-entrypoint-initwp.d/` directory check to allow users to mount and run arbitrary scripts on build.
- Use `twentynineteen` theme by default.

## 0.21.0

### Minor

- Automatically perform a WordPress database update after importing, if needed.

### Chore

- Update example files and error messages related to bindfs.

## 0.20.0

### Minor

- Add support for Redis object caching.
- Allow user to specify a different database charset via environment variable (#137)

## 0.19.0

**BREAKING CHANGE:** The WordPress service must now be configured in one of the following ways:

(preferred)
```yml
services:
    wordpress:
        cap_add:
            - SYS_ADMIN
        devices:
            - /dev/fuse
```

**OR**

(try to avoid if possible)
```yml
services:
    wordpress:
        privileged: true
```

This configuration allows the wordpress container to mount the WordPress
directory to the web root with back and forth syncronization using bindfs. This
configuration ensures that when editing the files/folders manually on your
system, they have the proper `admin` permissions in the /app directory, and
when the server edits files/folders from WordPress, the webroot always have the
proper `www-data` permissions in the web root.

## 0.18.0

**BREAKING CHANGE:** `URL_REPLACE` environment variable now only accepts `after_url`, rather than `before_url,after_url`. This will gracefully fix itself if it encounters the old format for this version only, but will break on subsequent versions.

## Minor
- Create and use a non-root user (admin) with passwordless sudo priviledges, rather than using the root user.
- Plugins and themes can now be space-separated (preferred) or comma-separated (legacy).
- Readdress #110 to print out logs in a similar, but greatly simplified, format.
- Install and remove themes and plugins in parallel.
- Greatly simplify build pipeline.

## 0.17.0

### Minor
- Add PHP 7.2 (`latest`, `latest-php7.2`, `0.17.0-php7.2`)
- Add `DB_USER` config option to environment variables

## 0.16.0
- **BREAKING CHANGE:** Builds will now exit if any plugin or theme installs fail.
- Improve logging.

## 0.15.2

### Fixes
- **REALLY** fix error causing volumed plugins and/or themes to be deleted.
- Move `.dockercache` out of index directory so it can't be accessed directly.

## 0.15.1

### Fixes
- Fix error causing volumed plugins to be deleted.
- Always flush `.htaccess` to add rewrite rules back in the event a plugin modified it (e.g. `w3-total-cache`).

### Other
- Add `wp-cli` bash completions to the wordpress container.

## 0.15.0

### Deprecations

**Note:** Both deprecations will still work as usual until the next release cycle.

#### `[local]plugin-name` & `[local]theme-name` syntax

The build process is now aware of locally-volumed plugins and themes automatically.

Additionally, listing locally-volumed plugins and themes in your `docker-compose.yml` file is optional; you may list them if you'd like to keep your compose file declarative, or you may skip listing them completely. The build will complete the same either way.

#### `SEARCH_REPLACE` has been renamed to `URL_REPLACE`

We chose to rename this because, although you may search and replace strings that are not URLs, the build process requires them to be.

This name reflects that requirement better and will lead to less confusion down the road.

### `VERBOSE` environement variable

Logging has been changed to show necessary information by default.

### Improvements
- Add `php7.1` base image.
- Widespread efficiency improvements to build process.
- Reduce the number of Dockerfile layers.
- If `SERVER_NAME` is specified (eg. `example.com`), create a `ServerAlias` in the apache configs for `www.example.com`.

### Fixes
- Plugins and themes are only pruned if they meet the following criteria (Closes #51):
  - Not listed in `docker-compose.yml`.
  - Not added as a local volume.
- Adjust permissions of volumed files and directories so that they remain editable outside the container, and remain secure. (Closes #12)
- Fix critical security vulnerability. HT @joerybruijntjes.

## 0.14.0
- Add `SERVER_NAME` variable to allow for quickly setting that directive in the apache config for users interested in running in production.

## 0.13.0
- Add `certbot` into the image to allow for easily obtaining and renewing SSL certificates

## 0.12.2
- Add missing `exif` php extension to handle WordPress media upload metadata parsing.
- Bump default base theme to `twentyseventeen`

## 0.12.1
- This project now maintains 2 separate `Dockerfile`, depending on the version of PHP you'd like to have installed.


run.sh
#!/bin/bash

if ! sudo mount -a 2> /dev/null; then
    printf '\e[1;31mERROR:\e[0m %s' \
        'Container running with improper privileges.

    Be sure your service is configured with the following options:
    ___
    services:
      wordpress:
        cap_add:
          - SYS_ADMIN
        devices:
          - /dev/fuse
        # needed on certain cloud hosts
        security_opt:
          - apparmor:unconfined
    ___

    OR (use first option if possible)
    ___
    services:
    wordpress:
        privileged: true
    ___

    ' | sed 's/^    //'
    exit 1
fi

# Environment
# ------------
declare -x TERM="${TERM:-xterm}"
declare PLUGINS="${PLUGINS//,/}"
declare THEMES="${THEMES//,/}"

# Runtime
# ------------
declare default_theme=twentytwenty
declare -A plugin_deps
declare -A theme_deps

# Configuration
# -------------
mkdir -p ~/.wp-cli
echo -e "
path: /app
color: true
apache_modules:
    - mod_rewrite

config create:
    dbhost: ${DB_HOST:-db}:3306
    dbname: ${DB_NAME:-wordpress}
    dbpass: ${DB_PASS:-root}
    dbprefix: ${DB_PREFIX:-wp_}
    dbuser: ${DB_USER:-root}
    dbcharset: ${DB_CHARSET:-utf8}
    extra-php: |
        define('WP_DEBUG', ${WP_DEBUG:-false});
        define('WP_DEBUG_LOG', ${WP_DEBUG_LOG:-false});
        define('WP_DEBUG_DISPLAY', ${WP_DEBUG_DISPLAY:-true});
        $(sed '1 ! s/.*/        \0/' < <(echo -e "${EXTRA_PHP:-}"))

core download:
    locale: ${WP_LOCALE:-en_US}
    skip-content: true
    version: ${WP_VERSION:-latest}

core install:
    admin_email: ${ADMIN_EMAIL:-admin@${SERVER_NAME:-localhost.com}}
    admin_password: ${DB_PASS:-root}
    admin_user: ${DB_USER:-root}
    skip-email: true
    title: ${DB_NAME:-wordpress}
    url: ${URL_REPLACE:-localhost:8080}

rewrite structure:
    hard: true

" > ~/.wp-cli/config.yml

main() {
    h1 'Begin WordPress Installation'
    init

    h2 'Waiting for MySQL to initialize...'
    while ! mysqladmin ping \
        --host="${DB_HOST:-db}" \
        --user="${DB_USER:-root}" \
        --password="${DB_PASS:-root}" \
        --silent > /dev/null; do
        sleep 1
    done

    h2 'Configuring WordPress'
    wp config create --force |& logger

    h2 'Checking database'
    check_database

    if [[ "$MULTISITE" == 'true' ]]; then
        h2 'Enabling Multisite'
        wp core multisite-convert |& logger
    fi

    h2 'Checking themes'
    check_themes

    h2 'Checking plugins'
    check_plugins

    h2 'Finalizing'
    if [[ $MULTISITE != true ]]; then
        wp rewrite structure \
            "${PERMALINKS:-/%year%/%monthnum%/%postname%/}" |& logger
    fi

    if [[ -e /docker-entrypoint-initwp.d ]]; then
        h2 'Executing user init scripts'
        for file in /docker-entrypoint-initwp.d/*; do
            [[ -x $file ]] && "$file"
        done
    fi

    h1 'WordPress Installation Complete!'

    sudo rm -f /var/run/apache2/apache2.pid
    sudo apache2-foreground
}

# Config Functions
# ---------------------

init() {
    declare raw_line
    declare -a keyvalue

    for raw_line in $PLUGINS; do
        mapfile -t keyvalue < <(
            sed -n '
                s/.*\[\(.*\)\]\([^[:blank:]]*\).*/\1\n\2/p # Matches [key]value form
                t                                          # If previous match succeeds, skip to end
                {p; p;}                                    # Assumes normal form
            ' <<< "$raw_line"
        )
        plugin_deps[${keyvalue[0]}]="${keyvalue[1]}"
    done

    for raw_line in $THEMES; do
        mapfile -t keyvalue < <(
            sed -n '
                s/.*\[\(.*\)\]\([^[:blank:]]*\).*/\1\n\2/p # Matches [key]value form
                t                                          # If previous match succeeds, skip to end
                {p; p;}                                    # Assumes normal form
            ' <<< "$raw_line"
        )
        theme_deps[${keyvalue[0]}]="${keyvalue[1]}"
    done

    # If no theme dependencies or volumes exist, fall back to default
    if [[ ${#theme_deps[@]} == 0 && ! -d /app/wp-content/themes ]]; then
        theme_deps["$default_theme"]="$default_theme"
    fi

    # Apache config adustments
    sudo sed -i \
        -e "/^[[:blank:]]*.ServerName www.example.com/{c\\" \
        -e "\\tServerName ${SERVER_NAME:-localhost} \\" \
        -e "\\tServerAlias www.${SERVER_NAME:-localhost}" \
        -e '}' /etc/apache2/sites-available/000-default.conf

    sudo chown -R admin:admin /app

    if [[ ! -f /app/wp-settings.php ]]; then
        h2 'Downloading WordPress'
        wp core download |& logger
    fi
}

# this is ran in a subshell to isolate nullglob as it for whatever reason
# interacts with our $PLUGINS and $THEMES variables.
# See: https://github.com/visiblevc/wordpress-starter/issues/176
check_database() (
    shopt -s nullglob
    declare file
    declare -i num_imported=0

    wp core is-installed 2> /dev/null && return

    wp db create |& logger

    wp core install |& logger

    for file in /data/*.sql; do
        h2 "Importing $file (this might take a while)..."
        wp db import "$file" |& logger
        ((num_imported++))
    done

    if ((num_imported > 0)); then
        if [[ -n $URL_REPLACE ]]; then
            h2 "Replacing URLs in database (this might take a while)..."
            wp search-replace \
                --skip-columns=guid \
                --report-changed-only \
                --no-report \
                "$(wp option get siteurl)" \
                "$URL_REPLACE" |& logger
        fi

        h2 'Updating database...'
        wp core update-db |& logger
    fi
)

check_plugins() {
    declare key

    for key in "${!plugin_deps[@]}"; do
        wp plugin is-installed "$key" || wp plugin install "${plugin_deps[$key]}" |& logger
        wp plugin is-active "$key" || wp plugin activate "$key" |& logger
    done
}

check_themes() {
    declare key

    for key in "${!theme_deps[@]}"; do
        wp theme is-installed "$key" || wp theme install "${theme_deps[$key]}" |& logger
        wp theme is-active "$key" || wp theme activate "$key" |& logger
    done
}

# Helpers
# ---------------------

declare -i term_width=70

h1() {
    declare border padding text
    border='\e[1;34m'"$(printf '=%.0s' $(seq 1 "$term_width"))"'\e[0m'
    padding="$(printf ' %.0s' $(seq 1 $(((term_width - $(wc -m <<< "$*")) / 2))))"
    text="\\e[1m$*\\e[0m"
    echo -e "$border"
    echo -e "${padding}${text}${padding}"
    echo -e "$border"
}

h2() {
    printf '\e[1;33m==>\e[37;1m %s\e[0m\n' "$*"
}

logger() {
    fold --width $((term_width - 9)) -s | sed -n '
    /^\x1b\[[0-9;]*m/{ # match any line beginning with colorized text
        s/Error:/  \0/ # pads line so its length matches others
        p              # any lines containing color
        b              # branch to end
    }
    s/.*/         \0/p # pads all other lines with 9 spaces
    '
}

main
